client/core/core.go

```diff
@@ -4973,11 +5023,16 @@ func handlePreimageRequest(c *Core, dc *dexConnection, msg *msgjson.Message) err
 	return nil
 }
 
-func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.OrderID) error {
+func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.OrderID, commitChecksum dex.Bytes) error {
 	tracker, preImg, isCancel := dc.findOrder(oid)
 	if tracker == nil {
 		return fmt.Errorf("no active order found for preimage request for %s", oid)
 	}
+
+	tracker.mtx.Lock()
+	tracker.csum = commitChecksum
+	tracker.mtx.Unlock()
```

Comment:
Changes to product code just to facilitate test, particularly blocking calls, is to be avoided like the plague.

If the test gets the `SubjectPreimageSent` note emitted below, we can be sure that this code has already executed and there's no risk of a data race.  Was a race detected?

Reply:
> Changes to product code just to facilitate tests

I agree, but this has nothing to do with tests, 

`tracker.csum` gets written to in one request (go-routine 1), while we a reading them in another request (go-routine 2),

maybe due to the time difference between these 2 requests it is safe to assume that `tracker.csum` will be synchronized, but its not 100% the case, so I'm just playing it safe here,
if this causes any performance issues, then we can probably omit locking here,

> Was a race detected?

nope, there is syncronization on channel that prevents race, given its current implementation

Reply:
We shouldn't roll the dice with data races based on general ideas about expected timing, no.

I'm commenting about the fact that in the path in `handleMatchProofMsg` you aren't locking the mutex for the `csum` read, only in the test and here in `processPreimageRequest`.  Hence it looks like you've set up the locking for the test.

But here's the sequence:
- when an epoch closes, server computes the csum from all order commitments in that epoch, and it sends preimage requests to clients for all those orders
- client gets a preimage request, starts `processPreimageRequest`, which records the csum that came with the request and **then** responds to the server with the preimage. So server only has the preimage if the csum is recorded.
- when the server gets all the epoch preimages (or times out on any of them), it performs matching with those it has preimages for
- server sends a match_proof note, which contains the preimages **that it received (obviously)** and the csum it computed from all the epoch order commitments
- `handleMatchProofMsg` looks (or should) for orders with the corresponding preimages.  If the a known preimage is found, that's because `processPreimageRequest` has completed already.

So if the preimage for the order is in the match_proof note, that means the server has received it from `processPreimageRequest` and the tracker.csum has already been saved.  No race.

Reply:
Okay, I see ... Turns out `preimage` and `match_proof` requests are both websocket requests that get through the same channel (thus it makes any inmemory state between them synchronized).

So, there is no need to synchronize explicitely.

Leaving here my flawed (due to lack of context) reasoning just for completeness.

<details>
<summary>my flawed reasoning</summary>
<p>

I understand the sequence you've described, but not sure what you mean by "saved" (`trackedTrade`s reside in process memory between `preimage` and `match_proof` requests) in "and the tracker.csum has already been saved. No race.".

> I'm commenting about the fact that in the path in handleMatchProofMsg you aren't locking the mutex for the csum read, only in the test and here in processPreimageRequest.

That's a very good point, and I didn't look at it from this perspective, let me clarify.

The way it is implemented right now, it looks indeed as if locking on `tracker.mtx` here is done purely for testing purposes. But it not the intent. 
The intent is - for any other go-rotine G that is interested in reading up-to-date `tracker.csum` value (or modifying it), the developer implementing G will only have to worry about aquiring/releasing this lock in G properly, and not follow through the code checking whether `tracker.csum` was written to with this lock being used or not.
This is the contract that, I assumed, everyone should follow when working with `trackedTrade` fields, protected by this lock (based on its description):
```go
	// mtx protects all read-write fields of the trackedTrade and the
	// matchTrackers in the matches map.
	mtx      sync.RWMutex
```
To summarize, we aquire this lock here in order to not shoot ourselves in the foot later on when implementing features that depend on `tracker.csum` value. 
Whether this opens a whole lot of different issues to worry about, such as deadlocks or bad performance, is another matter also worthy to think through.

Now, back to `handleMatchProofMsg`, why didn't I lock on `tracker.mtx` here ? This is my bad indeed, looks like we need to lock it there as well ...
</p>
</details>

Reply:
I'm honestly pleased that you're attuned to the need to synchronize with concurrent writes, but no need to go into it much - we're all on the same page about the need for that.

Yes, there is implicit synchronization that prevents concurrent writes, but it's not because "both websocket requests that get through the same channel" if I understand your meaning.  It's because a match_proof request can not include the preimage if `processPreimageRequest` has not already "saved" the csum (`tracker.csum = commitChecksum`). The other goroutine that reads `tracker.csum` will only attempt that read if the server knew about that preimage to include it with the match_proof note, which means it has already been stored in `tracker.csum`.  This all assumes the first iteration of these changes where preimages were checked to identify relevant trades.


Now a comment about the following:

> developer implementing G will only have to worry about aquiring/releasing this lock in G properly, and not follow through the code checking whether tracker.csum was written to with this lock being used or not.

You see this as being safe, but in reality not understanding access patterns and instead falling back to just locking mutexes liberally is seriously problematic.  When possible, it is generally wiser to design access patterns that do not involve the possibility for concurrent writes.  You'll have to believe me when I say it can become very messy with functions that maybe lock and maybe don't, called from other methods that do lock... so is it gonna deadlock or is it gonna be a race or is it gonna be correct?  This is particularly easy to get into with `trackedTrade.RWMutex` given how it's used so extensively in Core.

In short, the `csum` field should not be under the `mtx` in the struct definition because: it will only ever be accessed AFTER it is written, if the access patterns design for that.

Reply:
Although I've neglected the possibility of a malfunctioning server sending multiple preimage requests for the same order/commitment... :sigh:  I'm considerably less confident that we'd not end up with two concurrent writes (two `processPreimageRequest`).  We'd also want logic to reject such requests if `csum` is already stored.  Let's first see how this plays out after buck weighs in on my assertion that there can be epoch status orders for multiple epochs.

We might end up needing a sync mechanism for the `csum` field after all.  If that is the case we can consider it's own mutex or something more low-level like an atomic uint32 CAS that just says if it has been stored already so that we don't have to worry about an locking at the read sites.

Reply:
Thanks to the detailed explanations above, I think now we are on the same page regarding:
- related concurrency considerations
- we must assume the server is maliscious

Below are some conclusions I've arrived at, please let me know wheter any of these are incorrect.

1. Checking `len(trade.cSum) > 0` in `handleMatchProofMsg` is probably not sufficient, because maliscious server might just provide empty value for `CommitChecksum` in `handlePreimageRequest` resulting in client omitting `cSum` validation altogether.

2. I believe we shouldn't check for duplicate `handlePreimageRequest` requests from the server, instead we should apply every request (overwriting the previous `cSum` value) and just verifying the value matches should be enough to spot any maliscious behavior by the server. This looks like a simple enough solution, given (3).

3. 
> but that might warrant resetting it to 0 if the websocket Send just fails due to connectivity or whatever so a subsequent request can be fielded.

Not sure how to better handle that ... because it might be difficult to differentiate between "server playing games" & "connectivity issues".

Reply:
@chappjc @buck54321 any comments on my summary above ?

Reply:
> I believe we shouldn't check for duplicate handlePreimageRequest requests from the server, instead we should process every request (overwriting the previous csum value) and just verifying the value matches should be enough to spot any maliscious behavior by the server.

On the contrary, allowing duplicate requests to essentially update the commitment checksum is facilitating the server cheating.  Server would just have to send a final preimage request just before the match_proof ntfn with a csum that matches, thus defeating the check in handleMatchProof.

If the trade/cancel's `csum` is already set and not equal to what the repeated request is providing, that's definitely server cheating and should be handled as such.

Reply:
Also, if you could place a TODO comment in the test for handleMatchProof, we can get to that when we are ready to change these warnings into errors that prevent client from negotiating matches for such orders.

Reply:
Added a `TODO` comment about handling duplicates (just a comment, no code to handle that for now).

Reply:
I've approved the PR, but if @buck54321 has anything else that requires a commit, I'd really like to have `processPreimageRequest` return an error if the `tracker.{cancel.}csum` is non-nil.  Otherwise these changes offer no real protection yet, just a start in that direction.

Reply:
@chappjc I can create a follow-up PR to check for: `tracker.{cancel.}csum` is non-nil case.